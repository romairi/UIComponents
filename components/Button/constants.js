
import { SIZE_LARGE, SIZE_MEDIUM, SIZE_SMALL } from '../../constants/propTypes/sizes';

export const BUTTON_SIZES = {
    LARGE: SIZE_LARGE,
    MEDIUM: SIZE_MEDIUM,
    SMALL: SIZE_SMALL,
};
export const DEFAULT_SIZE = SIZE_MEDIUM;

export const BUTTON_TYPES = {
    BUTTON: 'button',
    RESET: 'reset',
    SUBMIT: 'submit'
};