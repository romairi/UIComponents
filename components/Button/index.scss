@import "../../style/colors/colors.scss";
@import "../../style/colors/functions.scss";

.r-button {
  cursor: pointer;
  padding: 10px 20px;
  font-size: 14px;
  color: white;
  background-color: get-primary($theme);
  border: none;
  border-radius: 7px;
  box-shadow: rgb(0 0 0 / 20%) 0px 3px 1px -2px,
    rgb(0 0 0 / 14%) 0px 2px 2px 0px, rgb(0 0 0 / 12%) 0px 1px 5px;
  transition: background-color 0.3s ease;

  &:hover {
    background-color: get-hover($theme);
  }

  &.small {
    font-size: 12px;
    padding: 8px 15px;
  }

  &.large {
    font-size: 18px;
    padding: 15px 24px;
  }

  &.secondary {
    background-color: get-secondary($theme);
    transition: background-color 0.5s ease;

    &:hover {
      background-color: get-hover-secondary($theme);
    }
  }

  &.error {
    background-color: get-error($theme);

    &:hover {
      background-color: get-hover-error($theme);
    }
  }

  &.text {
    color: get-hover($theme);
    background-color: white;
    box-shadow: none;
    border-radius: 0;
    transition: background-color 0.5s ease;

    &:hover {
      text-decoration: none;
      background-color: rgba(74, 101, 126, 0.08);
    }
  }
}
