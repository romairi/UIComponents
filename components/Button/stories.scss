.button-sizes {
  display: flex;
  justify-content: space-around;
  align-items: center;
}

.button-variants {
  display: flex;
  justify-content: space-around;
  align-items: center;
}
