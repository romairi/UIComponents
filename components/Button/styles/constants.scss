$boxShadowBtn: (
  rgb(0 0 0 / 20%) 0px 3px 1px -2px,
  rgb(0 0 0 / 14%) 0px 2px 2px 0px,
  rgb(0 0 0 / 12%) 0px 1px 5px
);
$textBackgroundBtn: (rgba(74, 101, 126, 0.08));
$outlinedBorder: (2px solid rgba(35, 120, 225, 1.5));

$button_size_mapping: (
  "small": (
    "font-size": 12px,
    "padding": 8px 15px
  ),
  "medium": (
    "font-size": 14px,
    "padding": 10px 20px
  ),
  "large": (
    "font-size": 18px,
    "padding": 10px 20px
  ),
);
