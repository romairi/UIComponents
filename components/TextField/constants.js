import {
  SIZE_LARGE,
  SIZE_MEDIUM,
  SIZE_SMALL,
} from "../../constants/propTypes/sizes";

export const TEXT_FIELD_VARIANTS = {
  FILLED: "filled",
  STANDARD: "standard",
  OUTLINED: "outlined",
};

export const TEXT_FIELD_SIZES = {
  LARGE: SIZE_LARGE,
  MEDIUM: SIZE_MEDIUM,
  SMALL: SIZE_SMALL,
};

export const TEXT_FIELD_TYPES = {
  TEXT: "text",
  BUTTON: "button",
  CHECKBOX: "checkbox",
  PASSWORD: "password",
  RADIO: "radio",
  RANGE: "range",
  SEARCH: "search",
};

export const DEFAULT_SIZE = SIZE_MEDIUM;
export const DEFAULT_VARIANT = TEXT_FIELD_VARIANTS.OUTLINED;
export const DEFAULT_TYPE = TEXT_FIELD_TYPES.TEXT;
