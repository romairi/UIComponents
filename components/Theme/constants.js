export const THEMES = {
    DEFAULT: 'default',
    ERROR: 'error',
};