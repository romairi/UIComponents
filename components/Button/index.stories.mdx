import { Meta, Story, Canvas } from "@storybook/addon-docs";
import Button from "./index";
import { BUTTON_SIZES, BUTTON_TYPES } from "./constants";

import "./stories.scss";

<Meta
  title="Button"
  component={Button}
  argTypes={{
    size: {
      options: Object.values(BUTTON_SIZES),
      control: { type: "select" },
    },
    type: {
      options: Object.values(BUTTON_TYPES),
      control: { type: "select" },
    },
  }}
/>

export const Template = ({
  primary,
  type,
  size,
  className,
  children = "SOME TEXT",
}) => (
  <Button className={className} type={type} size={size} primary={primary}>
    {children}
  </Button>
);

# Button

This is a building block component for a button.

## Default props

<Canvas>
  <Story name="default_props">{Template.bind({})}</Story>
</Canvas>

## Sizes

You can select different sizes for the button. Ones should be ["small", "medium", "large"]

<Canvas>
  <Story name="sizes">
    <div className="button_sizes">
      {Object.entries(BUTTON_SIZES).map(([size_name, size]) => (
        <Button key={size_name} size={size}>
          {size_name}
        </Button>
      ))}
    </div>
  </Story>
</Canvas>

<!-- ### small

<Canvas>
  <Story name="sizes_small">
    <Button size={BUTTON_SIZES.SMALL}>SOME TEST</Button>
  </Story>
</Canvas>

### medium

<Canvas>
  <Story name="sizes_medium">
    <Button size={BUTTON_SIZES.MEDIUM}>SOME TEST</Button>
  </Story>
</Canvas>

### large

<Canvas>
  <Story name="sizes_large">
    <Button size={BUTTON_SIZES.LARGE}>SOME TEST</Button>
  </Story>
</Canvas> -->
