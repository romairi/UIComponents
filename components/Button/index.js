import React from "react";
import _ from "lodash";
import PropTypes from "prop-types";
import { buildClassName } from "../../services/buildClassName";
import {
  BUTTON_SIZES,
  BUTTON_TYPES,
  BUTTON_VARIANTS,
  DEFAULT_SIZE,
  DEFAULT_VARIANT,
} from "./constants";
import Theme from "../Theme";

import "./index.scss";
import { IsTheme } from "../Theme/themePropTypes";

const propTypes = {
  /**
   * Is this the principal call to action on the page?
   */
  primary: PropTypes.bool,

  /**
   * The size of the button (small, medium, large).
   */
  size: PropTypes.oneOf(Object.values(BUTTON_SIZES)),

  /**
   * Button contents.
   */
  children: PropTypes.string.isRequired,

  /**
   * Optional click handler.
   */
  onClick: PropTypes.func,

  /**
   * The type of the button (button, reset, submit).
   */
  type: PropTypes.oneOf(Object.values(BUTTON_TYPES)),

  /**
   * ClassName to extend the component.
   */
  className: PropTypes.string,

  /**
   * The variant to use button (contained, text, outlined).
   */
  variant: PropTypes.oneOf(Object.values(BUTTON_VARIANTS)),

  /**
   * The theme of the button.
   */
  theme: IsTheme,
};

const defaultProps = {
  primary: true,
  size: DEFAULT_SIZE,
  type: BUTTON_TYPES.BUTTON,
  variant: DEFAULT_VARIANT,
  onClick: _.noop,
};

const Button = ({
  type,
  className,
  primary,
  size,
  variant,
  onClick,
  children,
  theme,
}) => {
  const mode = primary ? "" : "secondary";

  return (
    <Theme themeName={theme}>
      <button
        type={type}
        className={buildClassName([
          "r-button",
          className,
          mode,
          size,
          variant !== DEFAULT_VARIANT && variant,
        ])}
        onClick={onClick}
      >
        {children}
      </button>
    </Theme>
  );
};

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

export default Button;
