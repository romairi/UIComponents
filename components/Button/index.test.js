import React from "react";
import renderer from "react-test-renderer";
import Button from "./index";
import { BUTTON_SIZES, BUTTON_VARIANTS } from "./constants";
import { fireEvent, render } from "@testing-library/react";

describe("Button", () => {
  it("should render the default props", () => {
    const component = renderer.create(<Button>SOME TEXT</Button>);
    expect(component.toJSON()).toMatchSnapshot();
  });

  Object.entries(BUTTON_SIZES).forEach(([size_name, size]) => {
    it(`should render the button with the size: ${size_name}`, () => {
      const component = renderer.create(<Button size={size}>SOME TEXT</Button>);
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  it("should trigget the onClick callback", () => {
    const onClick = jest.fn();
    const { container } = render(<Button onClick={onClick}>SOME TEXT</Button>);

    fireEvent.click(container.firstChild.firstChild);
    expect(onClick).toHaveBeenCalled();
  });

  Object.values(BUTTON_VARIANTS).forEach((variant) => {
    it(`should render the button with the variant: ${variant}`, () => {
      const component = renderer.create(<Button variant={variant}>SOME TEXT</Button>);
      expect(component.toJSON()).toMatchSnapshot();
    });
  });
});
