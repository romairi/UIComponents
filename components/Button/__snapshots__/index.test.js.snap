// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render the button with the size: LARGE 1`] = `
<button
  className="r-button large"
  onClick={[Function]}
  type="button"
>
  SOME TEXT
</button>
`;

exports[`Button should render the button with the size: MEDIUM 1`] = `
<button
  className="r-button medium"
  onClick={[Function]}
  type="button"
>
  SOME TEXT
</button>
`;

exports[`Button should render the button with the size: SMALL 1`] = `
<button
  className="r-button small"
  onClick={[Function]}
  type="button"
>
  SOME TEXT
</button>
`;

exports[`Button should render the button with the variant: contained 1`] = `
<button
  className="r-button medium"
  onClick={[Function]}
  type="button"
>
  SOME TEXT
</button>
`;

exports[`Button should render the button with the variant: outlined 1`] = `
<button
  className="r-button medium outlined"
  onClick={[Function]}
  type="button"
>
  SOME TEXT
</button>
`;

exports[`Button should render the button with the variant: text 1`] = `
<button
  className="r-button medium text"
  onClick={[Function]}
  type="button"
>
  SOME TEXT
</button>
`;

exports[`Button should render the default props 1`] = `
<button
  className="r-button medium"
  onClick={[Function]}
  type="button"
>
  SOME TEXT
</button>
`;
